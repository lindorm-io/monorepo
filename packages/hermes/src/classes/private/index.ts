export * from "./HermesMetadata";
export * from "./HermesRegistry";
export * from "./HermesScanner";
