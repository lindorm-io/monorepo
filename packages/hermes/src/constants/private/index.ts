export * from "./checksum-store";
export * from "./domain";
export * from "./event-store";
export * from "./saga-store";
export * from "./view-store";
