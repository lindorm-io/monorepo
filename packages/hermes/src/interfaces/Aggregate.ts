import { ClassLike, Dict } from "@lindorm/types";
import { AggregateData } from "../types";
import { IHermesMessage } from "./HermesMessage";

export interface IAggregate<S extends Dict = Dict> extends AggregateData {
  apply(causation: IHermesMessage, event: ClassLike): Promise<void>;
  load(event: IHermesMessage): void;
  toJSON(): AggregateData<S>;
}
