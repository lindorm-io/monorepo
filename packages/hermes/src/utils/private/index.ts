export * from "./assert-checksum";
export * from "./compose-metadata";
export * from "./create-checksum";
export * from "./extract-data-transfer-object";
export * from "./get-view-store-name";
export * from "./global";
export * from "./recover-dto";
export * from "./recover-error";
export * from "./verify-name";
