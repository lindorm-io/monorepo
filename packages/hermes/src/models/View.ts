import { snakeCase } from "@lindorm/case";
import { ILogger } from "@lindorm/logger";
import { randomString } from "@lindorm/random";
import { DeepPartial, Dict } from "@lindorm/types";
import merge from "deepmerge";
import { z } from "zod";
import { ViewDestroyedError } from "../errors";
import { IView } from "../interfaces";
import { HermesEvent } from "../messages";
import { ViewData, ViewOptions } from "../types";
import { composeObjectMetadata } from "../utils/private/compose-metadata";

export class View<S extends Dict = Dict> implements IView<S> {
  public readonly id: string;
  public readonly name: string;
  public readonly context: string;

  private readonly _hash: string;
  private readonly _processedCausationIds: Array<string>;
  private readonly _revision: number;
  private _destroyed: boolean;
  private _meta: Record<string, any>;
  private _state: S;

  private readonly logger: ILogger;

  public constructor(options: ViewOptions<S>) {
    this.logger = options.logger.child(["View"]);

    this.id = options.id;
    this.name = snakeCase(options.name);
    this.context = snakeCase(options.context);

    this._destroyed = options.destroyed || false;
    this._hash = options.hash || randomString(16);
    this._meta = options.meta || {};
    this._processedCausationIds = options.processedCausationIds || [];
    this._revision = options.revision || 0;
    this._state = options.state || ({} as unknown as S);
  }

  // public properties

  public get destroyed(): boolean {
    return this._destroyed;
  }

  public get hash(): string {
    return this._hash;
  }

  public get meta(): Record<string, any> {
    return this._meta;
  }

  public get processedCausationIds(): Array<string> {
    return this._processedCausationIds;
  }

  public get revision(): number {
    return this._revision;
  }

  public get state(): S {
    return this._state;
  }

  // public

  public toJSON(): ViewData<S> {
    return {
      id: this.id,
      name: this.name,
      context: this.context,
      destroyed: this.destroyed,
      hash: this.hash,
      meta: this._meta,
      processedCausationIds: this.processedCausationIds,
      revision: this.revision,
      state: this.state,
    };
  }

  // public context

  public destroy(causation: HermesEvent): void {
    this.logger.debug("Destroy", { causation });

    if (this._destroyed) {
      throw new ViewDestroyedError();
    }

    this._destroyed = true;
  }

  public mergeState(causation: HermesEvent, data: DeepPartial<S>): void {
    this.logger.debug("Merge state", { data });

    z.record(z.any()).parse(data);

    if (this._destroyed) {
      throw new ViewDestroyedError();
    }

    const { state, meta } = composeObjectMetadata<S>(
      this._state,
      merge<any, any>(data, this._state),
      this._meta,
      causation.timestamp,
    );

    this._state = state;
    this._meta = meta;
  }

  public setState(causation: HermesEvent, data: S): void {
    this.logger.debug("Set state", { data });

    z.record(z.any()).parse(data);

    if (this._destroyed) {
      throw new ViewDestroyedError();
    }

    const { state, meta } = composeObjectMetadata<S>(
      this._state,
      data,
      this._meta,
      causation.timestamp,
    );

    this._state = state;
    this._meta = meta;
  }
}
