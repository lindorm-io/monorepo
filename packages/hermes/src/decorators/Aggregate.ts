import { Constructor } from "@lindorm/types";
import { AggregateDecoratorOptions } from "../types";
import { extractNameData, globalHermesMetadata } from "../utils/private";

export function Aggregate(options: AggregateDecoratorOptions = {}): ClassDecorator {
  return function (target) {
    const { name } = extractNameData(target.name);
    globalHermesMetadata.addAggregate({
      encryption: options.encryption || false,
      name: options.name || name,
      namespace: options.namespace || null,
      target: target as unknown as Constructor,
    });
  };
}
