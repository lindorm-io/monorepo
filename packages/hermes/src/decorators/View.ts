/* eslint-disable @typescript-eslint/no-unsafe-function-type */

import { isArray } from "@lindorm/is";
import { Constructor } from "@lindorm/types";
import { ViewDecoratorOptions, ViewStoreSource } from "../types";
import { extractNameData, globalHermesMetadata } from "../utils/private";

export function View(
  AggregateClass: Constructor | Array<Constructor>,
  source: ViewStoreSource,
  options: ViewDecoratorOptions = {},
): ClassDecorator {
  return function (target) {
    const { name } = extractNameData(target.name);
    globalHermesMetadata.addView({
      aggregates: isArray(AggregateClass) ? AggregateClass : [AggregateClass],
      name: options.name || name,
      namespace: options.namespace || null,
      source,
      target: target as unknown as Constructor,
    });
  };
}
