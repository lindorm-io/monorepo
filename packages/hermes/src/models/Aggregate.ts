import { snakeCase } from "@lindorm/case";
import { ILogger } from "@lindorm/logger";
import { ClassLike, Dict } from "@lindorm/types";
import merge from "deepmerge";
import { z } from "zod";
import {
  AggregateDestroyedError,
  AggregateNotDestroyedError,
  HandlerNotRegisteredError,
} from "../errors";
import { InvalidMessageTypeError } from "../errors/InvalidMessageTypeError";
import { HermesAggregateEventHandler } from "../handlers";
import { IAggregate, IHermesAggregateEventHandler, IHermesMessage } from "../interfaces";
import { HermesCommand, HermesEvent } from "../messages";
import { HermesMessageSchema } from "../schemas";
import { AggregateData, AggregateEventHandlerContext, AggregateOptions } from "../types";
import { extractDataTransferObject } from "../utils/private";

export class Aggregate<S extends Dict = Dict> implements IAggregate {
  public readonly id: string;
  public readonly name: string;
  public readonly context: string;

  private readonly _eventHandlers: Array<IHermesAggregateEventHandler>;
  private readonly _events: Array<IHermesMessage>;

  private _destroyed: boolean;
  private _destroying: boolean;
  private _numberOfLoadedEvents: number;
  private _state: S;

  private readonly logger: ILogger;

  public constructor(options: AggregateOptions) {
    this.logger = options.logger.child(["Aggregate"]);

    this.id = options.id;
    this.name = snakeCase(options.name);
    this.context = snakeCase(options.context);

    this._destroyed = false;
    this._destroying = false;
    this._eventHandlers = options.eventHandlers || [];
    this._events = [];
    this._numberOfLoadedEvents = 0;
    this._state = {} as unknown as S;
  }

  // public properties

  public get destroyed(): boolean {
    return this._destroyed;
  }

  public get events(): Array<IHermesMessage> {
    return this._events;
  }

  public get numberOfLoadedEvents(): number {
    return this._numberOfLoadedEvents;
  }

  public get state(): Record<string, any> {
    return this._state;
  }

  // public

  public async apply(causation: HermesCommand, event: ClassLike): Promise<void> {
    this.logger.debug("Apply Command", { causation, event });

    z.object({
      causation: z.object({ meta: z.record(z.any()) }),
      event: z.record(z.any()),
    }).parse({ causation, event });

    const { name, version, data } = extractDataTransferObject(event);

    await this.handleEvent(
      new HermesEvent(
        {
          aggregate: { id: this.id, name: this.name, context: this.context },
          data,
          name,
          meta: causation.meta,
          version,
        },
        causation,
      ),
    );
  }

  public async load(event: HermesEvent): Promise<void> {
    this.logger.debug("Load HermesEvent", { event });

    HermesMessageSchema.parse(event);

    await this.handleEvent(event);

    this._numberOfLoadedEvents += 1;
  }

  public toJSON(): AggregateData {
    return {
      id: this.id,
      name: this.name,
      context: this.context,
      destroyed: this.destroyed,
      events: this.events,
      numberOfLoadedEvents: this.numberOfLoadedEvents,
      state: this.state,
    };
  }

  // private

  private async handleEvent(event: HermesEvent): Promise<void> {
    this.logger.debug("Handle HermesEvent initialised", { event });

    try {
      if (!(event instanceof HermesEvent)) {
        throw new InvalidMessageTypeError(event, HermesEvent);
      }

      if (this._destroyed) {
        throw new AggregateDestroyedError();
      }

      const destroying = this._destroying;

      const eventHandler = this._eventHandlers.find(
        (x) => x.eventName === event.name && x.version === event.version,
      );

      if (!(eventHandler instanceof HermesAggregateEventHandler)) {
        this.logger.error("Event handler not registered", { event });
        throw new HandlerNotRegisteredError();
      }

      const ctx: AggregateEventHandlerContext = {
        event: structuredClone(event.data),
        logger: this.logger.child(["AggregateEventHandler"]),
        state: structuredClone(this.state),

        destroy: this.destroy.bind(this),
        destroyNext: this.destroyNext.bind(this),
        mergeState: this.mergeState.bind(this),
        setState: this.setState.bind(this),
      };

      await eventHandler.handler(ctx);

      if (destroying && !this._destroyed) {
        throw new AggregateNotDestroyedError();
      }

      this._events.push(event);

      this.logger.debug("Handle HermesEvent successful", { event });
    } catch (err: any) {
      this.logger.error("Handle HermesEvent failed", err);

      throw err;
    }
  }

  // private context

  private destroy(): void {
    this.logger.debug("Destroy");

    if (this._destroyed) {
      throw new AggregateDestroyedError();
    }

    this._destroyed = true;
    this._destroying = false;
  }

  private destroyNext(): void {
    this.logger.debug("Destroy Next");

    if (this._destroyed) {
      throw new AggregateDestroyedError();
    }

    this._destroying = true;
  }

  private mergeState(state: Partial<S>): void {
    this.logger.debug("Merge state", { state });

    z.record(z.any()).parse(state);

    if (this._destroyed) {
      throw new AggregateDestroyedError();
    }

    this._state = merge(this._state, state);
  }

  private setState(state: S): void {
    this.logger.debug("Set state", { state });

    z.record(z.any()).parse(state);

    if (this._destroyed) {
      throw new AggregateDestroyedError();
    }

    this._state = state;
  }
}
