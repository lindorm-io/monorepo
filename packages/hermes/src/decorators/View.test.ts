import { globalHermesMetadata } from "../utils/private";
import { Aggregate } from "./Aggregate";
import { View } from "./View";

describe("View Decorator", () => {
  @Aggregate()
  class TestViewAggregate {}

  test("should add metadata", () => {
    @View(TestViewAggregate, "mongo")
    class TestView {}

    expect(globalHermesMetadata.getView(TestView)).toMatchSnapshot();
  });

  test("should add metadata with custom options", () => {
    @View(TestViewAggregate, "mongo", {
      name: "custom_name",
      namespace: "custom_namespace",
    })
    class TestViewOptions {}

    expect(globalHermesMetadata.getView(TestViewOptions)).toMatchSnapshot();
  });
});
