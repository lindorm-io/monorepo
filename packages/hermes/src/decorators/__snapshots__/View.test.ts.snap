// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Decorator should add metadata 1`] = `
{
  "aggregates": [
    [Function],
  ],
  "handlers": [],
  "name": "test_view",
  "namespace": null,
  "source": "mongo",
  "target": [Function],
}
`;

exports[`View Decorator should add metadata with custom options 1`] = `
{
  "aggregates": [
    [Function],
  ],
  "handlers": [],
  "name": "custom_name",
  "namespace": "custom_namespace",
  "source": "mongo",
  "target": [Function],
}
`;
